### PR TITLE
t-rec 0.2.2 (new formula)

### PR DESCRIPTION
**Pull request by sassman**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by sassman** [comment]
tests are now green and this is the final version.

**Comment by gromgit** [comment]
If `t-rec` requires X to run, it should be explicitly declared as a dependency. To be sure, run:
```
brew linkage t-rec
```
If the output contains a section labeled **Undeclared dependencies with linkage**, you should add the formulae listed as dependencies. Then run `brew linkage t-rec` again and confirm that this section no longer appears.

**Comment by sassman** [comment]
It does not require X. It grabs screenshots from the Core graphics API. All dependencies are declared correctly. 
I consider the unknown headless environment on github workers to be the problem.

File: Formula/t-rec.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  def install
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    # let's fetch the window id
+    o = `#{bin}/t-rec -l | tail -1`
+    win_id = o.split(/\s|\n/)[-1]
+    # verify that it's an appropriate id
+    raise "No window id retrieved" unless win_id && Integer(win_id).positive?
```
Comment (SMillerDev) [inline Formula/t-rec.rb:20]: No need to raise in tests, please use assertions instead.

Comment (sassman) [reply Formula/t-rec.rb]: adjusted.

File: Formula/t-rec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    # let's fetch the window id
+    o = `#{bin}/t-rec -l | tail -1`
+    win_id = o.split(/\s|\n/)[-1]
+    # verify that it's an appropriate id
+    assert_equal win_id && Integer(win_id).positive?, true
+    # unfortunately grabbing a window screenshot does not work on github workers, so we cannot test this here
```
Comment (SMillerDev) [inline Formula/t-rec.rb:21]: Is there no command where you can actually check something about the output? This doesn't really say anything about the functionality of the program

Comment (sassman) [reply Formula/t-rec.rb:21]: Well as you can see in line 18 I split the output and verify that the second portion of a string in the format of `something | number` is a number. Furthermore, I'd like to point you to the [comment above](#issuecomment-730688428) where I'm trying to explain why a full featured test does not work on github actions, locally however it does work. The nature of a headless build server seems to be my guess of the culprit.

However, if you have specific suggestions feel free to propose them, I'm happy to implement them.

Comment (SMillerDev) [reply Formula/t-rec.rb:21]: What if you just pick an impossibly high window number and try and run the program on that? I expect it'll return some kind of error and you can check if the error code and text is as expected.

Comment (sassman) [reply Formula/t-rec.rb:21]: Sure, I can test that as well.

Comment (sassman) [reply Formula/t-rec.rb]: done

File: Formula/t-rec.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  test do
+    # let's fetch the window id
+    o = `#{bin}/t-rec -l | tail -1`
+    win_id = o.split(/\s|\n/)[-1]
+    # verify that it's an appropriate id
+    assert_equal win_id && Integer(win_id).positive?, true
+
+    # verify also error behaviour, as suggested
+    o = `WINDOWID=999999 #{bin}/t-rec 2>&1`
```
Comment (SMillerDev) [inline Formula/t-rec.rb:23]:
```suggestion
    o = shell_output("WINDOWID=999999 #{bin}/t-rec 2>&1")
```

File: Formula/t-rec.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  def install
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    # let's fetch the window id
+    o = `#{bin}/t-rec -l | tail -1`
+    win_id = o.split(/\s|\n/)[-1]
+    # verify that it's an appropriate id
+    assert_equal win_id && Integer(win_id).positive?, true
```
Comment (SMillerDev) [inline Formula/t-rec.rb:20]: This isn't needed as it doesn't test much.

Comment (sassman) [reply Formula/t-rec.rb:20]: Trust me on this, line 17 tests a lot of internal logic. I'd recommend leaving it in.

Comment (SMillerDev) [reply Formula/t-rec.rb:20]: In that case just do `shell_output("#{bin}/t-rec -l")`

File: Formula/t-rec.rb (hunk 3 above)
Comment (sassman) [reply Formula/t-rec.rb]: done

File: Formula/t-rec.rb (hunk 4 above)
Comment (sassman) [reply Formula/t-rec.rb]: done

**Comment by sassman** [comment]
Is there anything I can do, to help the PR pass?

**Review by chenrui333 — APPROVED** [review]
👍

**Comment by chenrui333** [comment]
@sassman, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Pull request closed without merging**